### PR TITLE
Add zoom.constrain.

### PR DESCRIPTION
**Pull request by mbostock**
Fixes #118. /cc @Herst

### CODE REVIEW

**Comment by Herst** [comment]
Thank you, I tested it and it works perfectly, with the following snippet I then achieved exactly what I wanted:
```js
.constrain(function (transform, extent, translateExtent) {
    var cx = transform.invertX((extent[1][0] - extent[0][0]) / 2),
        cy = transform.invertY((extent[1][1] - extent[0][1]) / 2),
        dcx0 = Math.min(0, cx - translateExtent[0][0]),
        dcx1 = Math.max(0, cx - translateExtent[1][0]),
        dcy0 = Math.min(0, cy - translateExtent[0][1]),
        dcy1 = Math.max(0, cy - translateExtent[1][1]);
    return transform.translate(
        Math.min(0, dcx0) || Math.max(0, dcx1),
        Math.min(0, dcy0) || Math.max(0, dcy1)
    );
})
```

On a side note, I never really worked with bundlers before (I hand-crafted shell scripts for my own projects :sweat_smile:) so it took my a bit of time and a question in d3's Slack channel to figure out what belongs to  `build/package.js` in order to get the standard `rollup -c` to work, I knew about <https://bl.ocks.org/mbostock/bb09af4c39c79cffcde4> but I just wanted to do a build as similar as possible to the official one (except for the `d3-zoom` package pointing to this PR branch of course), so the steps to go are the following:
```sh
git clone https://github.com/d3/d3.git
cd d3
npm install
# npm install git://github.com/d3/d3-SMTH.git#CUSTOM-BRANCH-OR-COMMIT --save
mkdir build
ver=$(git describe)
echo "export var version = '${ver:1}-CUSTOM';" > build/package.json
rollup -c
# result is in build/d3.js
```
Maybe this might be worth mentioning somewhere.

**Comment by Herst** [comment]
Here now also a demo with this in action: https://bl.ocks.org/Herst/093ff9962405dd564ef58ad8af9544d0

File: src/zoom.js [hunk 1]
```diff
@@ -38,17 +38,25 @@ function defaultTouchable() {
...
       wheelDelta = defaultWheelDelta,
       touchable = defaultTouchable,
-      k0 = 0,
-      k1 = Infinity,
-      x0 = -k1,
-      x1 = k1,
-      y0 = x0,
-      y1 = x1,
+      scaleExtent = [0, Infinity],
+      translateExtent = [[Infinity, Infinity], [-Infinity, -Infinity]],
```
Comment (Herst) [inline src/zoom.js]: Shouldn't that be `[[-Infinity, -Infinity], [Infinity, Infinity]]`?

**Comment by Herst** [comment]
On a side note, there is a little typo on the documentation page: "follwing".

File: src/zoom.js (hunk 1 above)
Comment (mbostock) [reply src/zoom.js]: Oops, yep!

**Pull request closed without merging**